### PR TITLE
Handle events

### PR DESCRIPTION
**Pull request by Antonhansel**

### CODE REVIEW

File: lib/helpers/get-events.js [hunk 1]
```diff
@@ -20,8 +36,30 @@ module.exports = function getEvents(serviceData, queues, cb) {
...
-      if(serviceData.next_stream_position) {
+      if(serviceData.next_stream_position && metadata.body.entries.length > 0) {
+        // Remove duplicates in Box response
+        // http://stackoverflow.com/questions/10429457/stream-position-returned-by-box-api-cannot-be-used-to-track-events
+        var stringEntries = [];
+        metadata.body.entries.forEach(function(item) {
+          stringEntries.push(JSON.stringify(item));
+        });
+        toUnique(stringEntries);
+        metadata.body.entries = [];
```
Comment (Neamar) [inline lib/helpers/get-events.js]: metadata.body.entries = stringEntries.map

File: lib/helpers/get-events.js [hunk 2]
```diff
@@ -20,8 +36,30 @@ module.exports = function getEvents(serviceData, queues, cb) {
         .end(cb);
     },
     function handleMetada(metadata, cb) {
+      var movedFoldersId = [];
       // if this is the first call, just store the new stream position
-      if(serviceData.next_stream_position) {
+      if(serviceData.next_stream_position && metadata.body.entries.length > 0) {
+        // Remove duplicates in Box response
+        // http://stackoverflow.com/questions/10429457/stream-position-returned-by-box-api-cannot-be-used-to-track-events
+        var stringEntries = [];
```
Comment (Neamar) [inline lib/helpers/get-events.js]: use .map

File: lib/helpers/get-events.js [hunk 3]
```diff
@@ -20,8 +36,30 @@ module.exports = function getEvents(serviceData, queues, cb) {
...
+        metadata.body.entries.forEach(function(item) {
+          stringEntries.push(JSON.stringify(item));
+        });
+        toUnique(stringEntries);
```
Comment (Neamar) [inline lib/helpers/get-events.js]: alright, let's remove this and send the duplicate twice

File: lib/helpers/get-events.js [hunk 4]
```diff
@@ -31,14 +69,41 @@ module.exports = function getEvents(serviceData, queues, cb) {
...
+          var crawler = new Crawler(serviceData, queues, 0, cb);
+          movedFoldersId.forEach(function(item) {
+            crawler.taskQueue.push(item);
+          });
+        }
       }
       serviceData.next_stream_position = metadata.body.next_stream_position;
-      cb(null);
+      // The crawler will use cb() if there was folders to index
+      if(movedFoldersId.length === 0) {
```
Comment (Neamar) [inline lib/helpers/get-events.js:74]: else

File: test/test-delete.js [hunk 5]
```diff
@@ -36,7 +36,7 @@ describe("Testing the delete event", function() {
     done();
   });
 
-  it('should require a new token, get the root folder and make the first /event call', function(done) {
+  it('Should add one file to the deletion queue', function(done) {
```
Comment (Neamar) [inline test/test-delete.js]: should (not Should)

File: test/test-duplicates.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    .reply(200, {
+      "next_stream_position": "1",
+      "entries": [{
+        "event_type": "ITEM_CREATE",
+        "source": {
+          "type": "file",
+          "id": "10"
+          }
+        },
+        {"event_type": "ITEM_CREATE",
```
Comment (Neamar) [inline test/test-duplicates.js]: {\n

File: test/test-duplicates.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    done();
+  });
+
+  it('should keep only one of the two events', function(done) {
+    update({refresh_token: 'fake_token', next_stream_position: '1'}, null, fakeQueue, function(err, cursor, serviceData) {
+      cursor.should.not.equal(null);
+      serviceData.access_token.should.equal('fake_access');
+      serviceData.next_stream_position.should.equal('1');
+      fakeQueue.addition.should.have.lengthOf(1);
+      fakeQueue.addition[0].id.should.equal('10');
```
Comment (Neamar) [inline test/test-duplicates.js]: `fakeQueue.addition[0].should.have.property('id', '10')`

File: test/test-duplicates.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    });
+
+    done();
+  });
+
+  it('should keep only one of the two events', function(done) {
+    update({refresh_token: 'fake_token', next_stream_position: '1'}, null, fakeQueue, function(err, cursor, serviceData) {
+      cursor.should.not.equal(null);
+      serviceData.access_token.should.equal('fake_access');
+      serviceData.next_stream_position.should.equal('1');
```
Comment (Neamar) [inline test/test-duplicates.js]: serviceData.should.have.property('next_stream_position', '1');

File: test/test-duplicates.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+      }]
+    });
+
+    done();
+  });
+
+  it('should keep only one of the two events', function(done) {
+    update({refresh_token: 'fake_token', next_stream_position: '1'}, null, fakeQueue, function(err, cursor, serviceData) {
+      cursor.should.not.equal(null);
+      serviceData.access_token.should.equal('fake_access');
```
Comment (Neamar) [inline test/test-duplicates.js]: should.have.property

File: test/test-move.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    .reply(200, {
+      "next_stream_position": "1",
+      "entries": [{
+        "event_type": "ITEM_MOVE",
+        "source": {
+          "type": "file",
+          "id": "10"
+          }
+        },
+        {"event_type": "ITEM_MOVE",
```
Comment (Neamar) [inline test/test-move.js]: {\n

File: test/test-move.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    done();
+  });
+
+  it('should add one file to the addition list and crawl the moved folder', function(done) {
+    update({refresh_token: 'fake_token', next_stream_position: '1'}, null, fakeQueue, function(err, cursor, serviceData) {
+      cursor.should.not.equal(null);
+      serviceData.access_token.should.equal('fake_access');
+      serviceData.next_stream_position.should.equal('1');
+      fakeQueue.addition.should.have.lengthOf(3);
+      fakeQueue.addition[0].id.should.equal('10');
```
Comment (Neamar) [inline test/test-move.js]: should.have.property

File: test/test-move.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+      fakeQueue.addition[1].id.should.equal('1');
```
Comment (Neamar) [inline test/test-move.js]: should.have.property

File: test/test-move.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+      fakeQueue.addition[2].id.should.equal('2');
```
Comment (Neamar) [inline test/test-move.js]: should.have.property

File: lib/helpers/get-events.js (hunk 4 above)
Comment (Antonhansel) [inline lib/helpers/get-events.js:74]: No, the cb only needs to be called if there was no new folders.

File: test/test-create.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    .reply(200, {
+      "next_stream_position": "1",
+      "entries": [{
+        "event_type": "ITEM_CREATE",
+        "source": {
+          "type": "file",
+          "id": "10"
+          }
+        },
+        {"event_type": "ITEM_UPLOAD",
```
Comment (Neamar) [inline test/test-create.js]: {\n

**Pull request merged**